packages/grid/x-data-grid/src/hooks/features/columns/useGridColumnGrouping.ts

```diff
@@ -0,0 +1,212 @@
+    }
+    groupLookup = { ...groupLookup, ...subTreeLookup, [groupId]: groupParam };
+  });
+
+  return { ...groupLookup };
+};
+export const columnGroupsStateInitializer: GridStateInitializer<
+  Pick<DataGridProcessedProps, 'columnGroupingModel'>
+> = (state, props) => {
+  const groupLookup = createGroupLookup(props.columnGroupingModel ?? []);
+  return {
+    ...state,
+    columnGrouping: { lookup: groupLookup, groupCollapsedModel: {} },
+  };
+};
+
+/**
+ * @requires useGridColumns (method, event)
+ * @requires useGridParamsApi (method)
+ */
+export const useGridColumnGrouping = (
+  apiRef: React.MutableRefObject<GridApiCommunity>,
+  props: Pick<DataGridProcessedProps, 'columnGroupingModel'>,
+) => {
+  /**
+   * API METHODS
+   */
+  const getColumnGroupPath = React.useCallback<GridColumnGroupingApi['getColumnGroupPath']>(
+    (field) => {
+      const columnLookup = gridColumnLookupSelector(apiRef);
+
+      return columnLookup[field]?.groupPath ?? [];
+    },
+    [apiRef],
+  );
+
+  const getAllGroupDetails = React.useCallback<GridColumnGroupingApi['getAllGroupDetails']>(() => {
+    const columnGroupLookup = gridColumnGroupsLookupSelector(apiRef);
+    return columnGroupLookup;
+  }, [apiRef]);
+
+  const getGroupDetails = React.useCallback<GridColumnGroupingApi['getGroupDetails']>(
+    (groupId) => {
+      const columnGroupLookup = gridColumnGroupsLookupSelector(apiRef);
+
+      return columnGroupLookup[groupId] ?? null;
+    },
+    [apiRef],
+  );
+
+  const columnGroupingApi: GridColumnGroupingApi = {
+    getColumnGroupPath,
+    getGroupDetails,
+    getAllGroupDetails,
+  };
+
+  useGridApiMethod(apiRef, columnGroupingApi, 'GridColumnGroupingApi');
+
+  /**
+   * Processors
+   * */
+  const addHeaderGroups = React.useCallback<GridPipeProcessor<'hydrateColumns'>>(
+    (columnsState) => {
+      if (!props.columnGroupingModel) {
+        return columnsState;
+      }
+      const unwrappedGroupingModel = unwrapGroupingColumnModel(props.columnGroupingModel);
+      if (Object.keys(unwrappedGroupingModel).length === 0) {
+        return columnsState;
+      }
+
+      columnsState.all.forEach((field) => {
+        const newGroupPath = unwrappedGroupingModel[field] ?? [];
+
+        const lookupElement = columnsState.lookup[field];
+        if (hasGroupPath(lookupElement) && isDeepEqual(newGroupPath, lookupElement?.groupPath)) {
+          // Avoid modifying the pointer to allow shadow comparison in https://github.com/mui/mui-x/blob/f90afbf10a1264ee8b453d7549dd7cdd6110a4ed/packages/grid/x-data-grid/src/hooks/features/columns/gridColumnsUtils.ts#L446:L453
+          return;
+        }
+        columnsState.lookup[field] = {
+          ...columnsState.lookup[field],
+          groupPath: unwrappedGroupingModel[field] ?? [],
+        };
+      });
+      return columnsState;
+    },
+    [props.columnGroupingModel],
+  );
+
+  useGridRegisterPipeProcessor(apiRef, 'hydrateColumns', addHeaderGroups);
```

Comment:
Could you extract the pre-processsor to another hook and initialize it before all other columns pre-processors? When the pre-processors for `hydrateColumns` run for the first time, all other pre-processor should also have been initialized to avoid unnecessary rerenders. 

Take as example `useGridSelectionPreProcessors`: https://github.com/mui/mui-x/blob/master/packages/grid/x-data-grid/src/hooks/features/selection/useGridSelectionPreProcessors.ts

https://github.com/mui/mui-x/blob/3b75125cab5c6dc70cf24b966c6d082b81d6a7f7/packages/grid/x-data-grid/src/DataGrid/useDataGridComponent.tsx#L55-L56